client/cmd/dexc/ui/depthchart.go

```diff
@@ -0,0 +1,279 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package ui
+
+import (
+	"fmt"
+	"math"
+	"sync"
+
+	"github.com/gdamore/tcell"
+	"github.com/rivo/tview"
+)
+
+var (
+	chartColor = tcell.GetColor("#0c456b")
+)
+
+type depthPoint struct {
+	x float64
+	y float64
+}
+
+// depthChart is a simple character-based chart. Right now, the chart only
+// works for a fixed chart size which cannot be changed.
+type depthChart struct {
+	*tview.Box
+	focus bool
+	// The chart data is protected by a mutex since more than one thread could be
+	// accessing simultaneously. That is not the case for focus, which is only
+	// accessed during calls initiated by tview, so assumed to be sequenced.
+	mtx      sync.RWMutex
+	seq      uint64
+	drawID   uint64
+	runeRows []string
+}
+
+func newDepthChart() *depthChart {
+	dc := &depthChart{
+		Box: tview.NewBox(),
+		seq: 1,
+	}
+	dc.SetBorder(true)
+	dc.runeRows = dc.calcRows()
+	return dc
+}
+
+// Need to handle concurrency for chart updates.
+func (c *depthChart) newScreenRows() []string {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	if !c.focus {
+		return nil
+	}
+	if c.seq == c.drawID {
+		// Nothing to redraw.
+		return nil
+	}
+	c.runeRows = c.calcRows()
+	c.seq = c.drawID
+	return c.runeRows
+}
+
+// Draw hijacks the embedded Box method to draw on the screen.
+func (c *depthChart) Draw(screen tcell.Screen) {
+	screenX, screenY, width, _ := c.GetRect()
+	rows := c.newScreenRows()
+	for y, row := range rows {
+		tview.Print(screen, row, 0+screenX, y+screenY, width, tview.AlignLeft, chartColor)
+	}
+}
+
+// refresh recalculates the chart rune rows.
+//
+// DRAFT NOTE: This method is currently just a demo of charting. It will need to
+// be a lot smarter when real data is plotted.
+func (c *depthChart) calcRows() []string {
+	rotations := float64(2.5)
+	numPts := 400
+	amp := float64(5)
+	pts := make([]depthPoint, 0, numPts)
+	for i := 0; i < numPts; i++ {
+		x := float64(i) / float64(numPts) * rotations * 2 * math.Pi
+		pts = append(pts, depthPoint{
+			x: float64(i),
+			y: math.Sin(x) * amp,
+		})
+	}
+	width, height := marketChartWidth, marketChartHeight
+	if width == 0 || height == 0 {
+		return nil
+	}
+	runeRows := make([]string, 0, height)
+	edge, err := interpolate(pts, width, height)
+	if err != nil {
+		// Don't update the UI from a Draw method, so we won't print an error here.
+		// TODO: Add a file-only logger for logging these errors?
+		return nil
```

Comment:
Seems like we need a way to get errors back to the application log or notifications window.  That can be for another day though.  I have no idea how simple that would be to do elegantly.

Reply:
Fixed in ba8ce3d